### ISSUE TITLE
Multivariate-Tweedie for age/length compositional likelihood

### ISSUE THREAD

**Issue by James-Thorson-NOAA** [nmfs-stock-synthesis/stock-synthesis#266]
I'm leading a paper with Tim Miller and Brian Stock, introducing a new "multivariate-Tweedie" likelihood for fitting age/length composition data.  It is essentially a more flexible alternative to the Dirichlet-multinomial, and appears to have some useful pros and cons relative to the DM.  It internally relies upon a `dtweedie` likeliood, which Johnoel is adding the ADMB-13, so it should be feasible to implement in SS3 by this summer.

Does anyone have sufficient time and interest to work with me to get it implemented (both in logical code, and with a thoughtful user-interface) in SS3?

**Comment by iantaylor-NOAA** [comment]
I don't think I can do this without help from others on the SS3 team, but am happy to be involved.

**Comment by James-Thorson-NOAA** [comment]
yeah, it'll have two parameters instead of one ... if someone could make a dev branch or something with those named parameters, I could do a PR with the added code.  Or is there some other preferred way to proceed ... perhaps instead I could just providing the code-snippet for the likelihood in this Issue thread?

**Comment by Rick-Methot-NOAA** [comment]
I did not intend to fast-track this.  I just wanted there to be a branch on which development could happen.  Only git collision I foresee is with the request to rename the D-M parameters.

**Comment by James-Thorson-NOAA** [comment]
By `read_data` I assume you mean "SS_readdata_330.tpl" ... I just opened this and can't easily make sense of it.  So a few questions:

1. Would it still be helpful if I can make the changes in `SS_objfun` but not `SS_readdata_330`?
2. If yes to Q1, could someone provide me with syntax to extract the parameters from the various inputs, e.g., analogous to how the DM does is using lines:

```R
        if(Comp_Err_L(f)==1) dirichlet_Parm=mfexp(selparm(Comp_Err_Parm_Start+Comp_Err_L2(f)))*nsamp_l(f,i);
        if(Comp_Err_L(f)==2) dirichlet_Parm=mfexp(selparm(Comp_Err_Parm_Start+Comp_Err_L2(f)));
```

...?

**Comment by Rick-Methot-NOAA** [comment]
Jim,
If you can start by working in SS_objfun.tpl.  When you push that, I can see what parameters need to be created in SS_readdata330.tpl, and then named in ss_readcontrol330.tpl.  Everything else should be handled automatically (reporting and write_ssnew).

**Comment by iantaylor-NOAA** [comment]
The potential conflict noted by Rick is with issue #185 where there's a proposal to add information about which fleets use the parameter to the label, such as "ln(DM_theta)_Age_P7(2-3-4)" instead of the status-quo "ln(DM_theta)_7" for the 7th DM parameter. However, that change is low priority so can wait until MV-Tweedie is developed in parallel with some similar naming convention.

Is it possible, and would it ever be a good idea, to share one of these weighting parameters among a mix of length and age data? If so, we should dial back the proposal to add "Age" to the label.

**Comment by iantaylor-NOAA** [comment]
Jim, nice job getting this together. In addition to the additional work that Jim has noted, we will need to install the `dev` branch of ADMB to get the `dtweedie` function https://github.com/admb-project/admb/blob/dev/src/linad99/dtweedie.cpp (which indeed seems to not be vectorized).

I'm too tired from a week of workshop to do any real work on this but I'll make a to-do list below to keep track of the next steps.

@k-doering-NOAA, it's probably a good idea to be testing ADMB 13 prior to it's release for 3.30.19 to make sure there are no compile errors or changes in model results. Would it be possible to add a github action to do so?

To-do list for MV Tweedie likelihood based on comments above:
- [x] create new [MV-Tweedie-complike](https://github.com/nmfs-stock-synthesis/stock-synthesis/tree/MV-Tweedie-complike) branch to host changes code
- [x] add new likelihood calculation to `SS_objfun.tpl` and create a pull request (#278) 
- [ ] create github action to compile executable using `dev` branch of ADMB
- [ ] double check likelihood calculation added to `SS_objfun.tpl` by @James-Thorson-NOAA 
- [ ] see what parameters need to be created in `SS_readdata_330.tpl`
- [ ] name parameters in `SS_readcontrol_330.tpl` (two for mvtweedie compared with one for DM)
- [ ] check everything else is be handled automatically (reporting and write_ssnew)
- [ ] document the changes in the SS3 User Manual
- [ ] apply any parallel changes to DM likelihood (discussed in #185 and #219)
- [ ] modify r4ss to work with models that use MV Tweedie

**Comment by Rick-Methot-NOAA** [comment]
Nice work Ian and Jim.  
I will work on the readdata and readcontrol aspects.
For #219 , I intend to look into modularizing this convoluted section of code that has lots of repeated elements.

**Comment by iantaylor-NOAA** [comment]
Rick, that sounds great. 

Question on the workflow for a change like this (which I just assigned to the 3.30.20 milestone): there's a pull request associated with this issue (#277 from `MV-Tweedie-complike` into `main`). We could either leave that open until all this is done, or just create the PR once everything is in place. 

@k-doering-NOAA, does leaving the PR open trigger more checks, or do we get those for the branch regardless? 
An open PR also gives people notifications when changes are pushed to that branch, which could be either good or bad depending on how many notifications we all get.

**Comment by Rick-Methot-NOAA** [comment]
Didn't realize that Jim was not yet part of the SS team, so just invited him into NOAA_contributors.
Looked at the parameter I/O.  This will be a bit complicated because current logic of counting parameters is completely dependent on one parameter and does not generalize easily to 2 parameters.  so, this will take a bit of time.

**Comment by iantaylor-NOAA** [comment]
Good solution to add Jim to the contributors who can push changes within the organization. 
I think we can probably close the pull request and all contribute to the new branch until it's ready, then open a new PR. 
@k-doering-NOAA may have additional ideas on the optimal workflow when she logs in tomorrow.

**Comment by k-doering-NOAA** [comment]
Wow, lots of action on this topic, this is great! A few thoughts.
1. Agree it would be good to test ADMB 13 separately. Since it is not out yet, I don't think we should be using it for the 3.30.19 release, but I can set up a job that runs the dev version of admb for testing out MV tweedie changes.
2. No need to leave the PR open, as I think all of the jobs should run when pushing to the branch anyway. We could close it, or mark the PR as "draft" to indicate it is not yet done. Either way works!

**Comment by k-doering-NOAA** [comment]
Just wanted to say, setting up the GHA to test out admb dev is on my to do list! However, we are working on release 3.30.19, so that will come first. I hope to get to the admb dev by next week.

**Comment by Rick-Methot-NOAA** [comment]
But we still could not continue development of MVT because any commit would trigger a gha which would fail when it tried to compile the MVT command.

**Comment by k-doering-NOAA** [comment]
I'm confused, could the failing jobs on the branch just be ignored for now? I think we would want them to pass before merging the branch into main, but I'm hoping ADMB 13 will be out before then, so I can modify our testing suite to use it.

**Comment by Rick-Methot-NOAA** [comment]
What I mean is that we cannot proceed with development of a branch that includes a call like Do_MVT().  Even though we could compile SS3 locally using ADMB13, compiles using gha will fail because the do_MVT() routine does not exist in ADMB 12.3

**Comment by k-doering-NOAA** [comment]
I get that the do_MVT() routine doesn't exist in ADMB 12.3, so all the github actions would fail. I guess I still don't understand why that would impede development?

**Comment by Rick-Methot-NOAA** [comment]
I suppose we still could do all MVT development locally, but we created gha partly so we could test frequently

**Comment by k-doering-NOAA** [comment]
Ah, ok, I see the concern. Maybe it would be best to wait on this until ADMB 13 (or its prerelease) comes out

**Comment by James-Thorson-NOAA** [comment]
Hi all,

I don't know where this landed, but I saw that ADMB-13 is now released, and also the MVTW paper is now accepted @ ICESJMS.  

Is there anything I can do, or perhaps you could tell me when you've started switching over to ADMB-13?

Jim

**Comment by Rick-Methot-NOAA** [comment]
There is a branch with a PR in which a placeholder to Tweedie is implemented, as well as several changes to how D-M is implemented.
We have installed ADMB13 locally and tested SS3 with it.  All looks good.
I'll get back to you in a week with an update on when we are ready to move to next stage of implementing Tweedie.

**Comment by James-Thorson-NOAA** [comment]
Kathryn or Ian,

can either of you (or someone else) point me at the SS3 code snippet where the logical code for the mvtweedie would be added (presumably by me)?  

I'm happy to take a stab at this in the next couple weeks.

Jim

**Comment by iantaylor-NOAA** [comment]
Hi @James-Thorson-NOAA,
The length comp likelihood for the MV-tweedie is implemented here (thanks to you): https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/main/SS_objfunc.tpl#L376-L401. There are placeholders in that file for the age and generalized size-comp likelihoods. I'm assuming that @Rick-Methot-NOAA is planning to move the likelihood into a generalized function for all data types in this section: https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/main/SS_miscfxn.tpl#L106-L122 (where the multinomial and Dirichlet-multinomial are now located).

I'm not sure what other changes are needed to implement this. I checked off the items for the data and control file pieces in the checklist at the top of this issue as it looks like they are complete.

**Comment by Rick-Methot-NOAA** [comment]
Ian covered the new situation well.  Suggest creating a new branch if you work on the full Tweedie implementation.  Adding Elizabeth for awareness.  @e-gugliotti-NOAA

**Comment by Rick-Methot-NOAA** [comment]
The Tweedie code will move into ss_miscfxn.tpl, similar to the function:  Comp_logL_Dirichlet()

**Comment by James-Thorson-NOAA** [comment]
@Rick-Methot-NOAA -- I just checked the `main` and `MV-Tweedie-complike` branches of `stock-synthesis` repo, looked in `ss_miscfxn.tpl` and don't see `Comp_logL_Dirichlet`.

I imagine I'd need more help to make any further progress, and will take this off my short-term to-do list for now.  But I'm happy to jump back in once I understand the plan for moving forward.

**Comment by Rick-Methot-NOAA** [comment]
I will finish the implementation, then pass to Jim for testing.

The code is in main and I will work from there.

The FUNCTION in ss_miscfxn.tpl is:
FUNCTION dvariable Comp_logL_Dirichlet(const double& Nsamp, const dvariable& dirichlet_Parm, const dvector& obs_comp, const dvar_vector& exp_comp)
  {
    dvariable logL;
    logL = sum(gammln(Nsamp * obs_comp + dirichlet_Parm * exp_comp)) - sum(gammln(dirichlet_Parm * exp_comp));
    return (logL);
  }

**Comment by James-Thorson-NOAA** [comment]
OK thanks Rick!  

PS:  I posted a comment where I was confused about where to find files.  I then deleted the comment because I realized it was due to working on my own fork which was out-of-date.  I'm definitely struggling a bit to sort through the different file structures across branches and forks :0
